### ISSUE TITLE
Problems with semi, semi-spacing, and no-extra-semi all turned on

### ISSUE THREAD

**Issue by dotnil** [eslint/eslint#2837]
Setting:

```js
{
  "rules": {
    "semi": [2, "never"],
    "semi-spacing": [2, { "before": false, "after": true }],    // default
    "no-extra-semi": 2
}
```

The following patterns are considered warning:

```js
;(typeof console === 'undefined' ? alert : console.log)(/* ... */)
```

because theres no space after the leading semicolon. But I think the style given in this example shall be the preferred one.

The following patterns are considered warning too:

```js
function foo() {}
;(foo() ? alert : console.log)(/* ... */)
```

because `no-extra-semi` rule is applied. eslint will consider the leading semicolon unnecessary. But this semicolon is vital...

**Comment by michaelficarra** [comment]
What makes you think that that semicolon is vital? Have you tried removing it?

**Comment by nzakas** [comment]
Duplicate of #1983

**Comment by nzakas** [comment]
Oops, the semi-spacing piece is the dupe. The semicolon in your last example is unnecessary, which is why no-extra-semi warns.

**Comment by dotnil** [comment]
@michaelficarra @nzakas aha my mistake. I thought it would be executed as below if the leading semicolon is omitted:

```js
function foo() {}(foo() ? alert : console.log)(/* ... */)
```

Thanks for the correction.
